File: vision/retina_parity.py
#!/usr/bin/env python3
"""视网膜采样的 parity 参考：Python 采一次，JS 采一次，必须逐值对上。

测试画面用**纯整数算式**生成（i*2654435761 % 65536），两边逐位一致，
这样比对失败一定是采样逻辑的问题，不会是浮点噪声。

用法：conda activate fba && python vision/retina_parity.py
"""
import json
from pathlib import Path
import numpy as np

ROOT = Path(__file__).resolve().parent.parent
OUT = ROOT / "results/vision"


def main():
    meta = json.loads((OUT / "retina.json").read_text())
    w, h = meta["width"], meta["height"]
    idmap = np.frombuffer((OUT / "retina_map.bin").read_bytes(), "<u2").astype(int)
    assert idmap.size == w * h, (idmap.size, w, h)
    npx = np.asarray(meta["pixels_per_ommatidium"], float)
    perm = np.asarray(meta["flygym_to_flyvis"], int)

    i = np.arange(w * h, dtype=np.int64)
    frame = ((i * 2654435761) % 65536) / 65536.0        # 整数 → 精确

    s = np.bincount(idmap, weights=frame, minlength=722)[1:]
    gym = s / npx
    hexv = gym[perm]

    doc = dict(width=w, height=h,
               frame_formula="frame[i] = ((i*2654435761) % 65536) / 65536",
               hex=np.round(hexv, 12).tolist())
    f = OUT / "retina_parity.json"
    f.write_text(json.dumps(doc, separators=(",", ":")))
    print(f"写入 {f}  {f.stat().st_size/1024:.0f} KB")
    print(f"  画面 {w}×{h}，721 个小眼亮度 范围 [{hexv.min():.6f}, {hexv.max():.6f}] 均值 {hexv.mean():.6f}")


if __name__ == "__main__":
    main()

File: vision/retina_parity.js
#!/usr/bin/env node
/** 视网膜采样：JS 必须和 Python 逐值对上。 */
const fs = require("fs"), path = require("path");
const Retina = require("./retina.js");
const ROOT = path.resolve(__dirname, "..");
const P = f => path.join(ROOT, "results/vision", f);

for (const f of ["retina.json", "retina_map.bin", "retina_parity.json"]) {
  if (!fs.existsSync(P(f))) {
    console.error(`缺少 ${P(f)}\n先跑：conda activate fba && python vision/export_retina_js.py && python vision/retina_parity.py`);
    process.exit(1);
  }
}
const meta = JSON.parse(fs.readFileSync(P("retina.json"), "utf8"));
const ref = JSON.parse(fs.readFileSync(P("retina_parity.json"), "utf8"));
const r = Retina.fromBuffer(fs.readFileSync(P("retina_map.bin")), meta);

const n = r.w * r.h;
const frame = new Float64Array(n);
for (let i = 0; i < n; i++) frame[i] = ((i * 2654435761) % 65536) / 65536;
const got = r.sample(frame);

let mx = 0, at = -1;
for (let j = 0; j < got.length; j++) {
  const d = Math.abs(got[j] - ref.hex[j]);
  if (d > mx) { mx = d; at = j; }
}
console.log(`画面 ${r.w}×${r.h} → ${got.length} 个小眼`);
console.log(`最大绝对差 ${mx.toExponential(2)}（第 ${at} 柱）`);
if (mx < 1e-12) {
  console.log("✓ 视网膜采样与 Python 完全一致");
} else {
  console.log("✗ 对不上 —— 检查 idmap 解析、每小眼像素数、flygym→flyvis 排列");
  process.exit(1);
}
